docs/spec_draft.md

```diff
@@ -1267,6 +1268,70 @@ More formally, `result[i0, ..., iR-1]` is equal to:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reduce
+
+### Semantics
+
+Applies a `body` function to `inputs` and `init_values` along the `dimensions`
+and produces a `result` tensor.
+
+The `body` function has to be associative, and it produces an
+implementation-defined behavior otherwise. Similarly, if `init_values` do not
+form identity under the `body` function, it produces an implementation-defined
+behavior.
+
+### Inputs
+
+| Name          | Type                                             |
+|---------------|--------------------------------------------------|
+| `inputs`      | variadic number of tensors of any supported type |
+| `init_values` | variadic number of scalars of any supported type |
+| `dimensions`  | 1-dimensional tensor constant of type `si64`     |
+| `body`        | `function`                                       |
+
+### Outputs
+
+| Name      | Type                                             |
+|-----------|--------------------------------------------------|
+| `results` | variadic number of tensors of any supported type |
+
+### Constraints
+
+  * (C1) `inputs`, `init_values` have the same element type.
+  * (C2) All `inputs` have the same shape.
+  * (C3) type(`inputs[k]`) $=$ type(`init_values[k]`) for all `k` $\in$ [0, N).
+  * (C4) size(`inputs`) $=$ size(`init_values`) $=$ size(`result`) $=$ N where
+  N >= 1.
+  * (C5) `init_values` have to form an identity under the `body`.
+  * (C6) 0 $\lt$ `dimensions[d]` $\lt$ rank(`inputs[d]`) for all dimension `d`.
+  * (C7) All dimensions in `dimensions` are unique.
+  * (C8) The full function type of `body` is
+  `(T0,..., TN-1, T0, ..., TN-1) -> (T0, ..., TN-1)`, where the first and second
+  set of input types `T0,..., TN-1` corresponds to the element types of `inputs`
+  and `init_values` respectively. The output types `(T0,..., TN-1)` corresponds
+  to the element types of `results`.
```

Comment:
```
* (C8) `body` has type `(tensor<E0>, ..., tensor<EN-1>, tensor<E0>, ..., tensor<EN-1>) -> (tensor<E0>, ..., tensor<EN-1>)`
where `Ek = element_type(inputs[k])`.
```

Feedback:
  1. I don't think constraints should have an opinion on whether the first or the second half of parameters map on inputs or init_values, unless this matters for verification / type inference. This should be covered in semantics.
  2. `Ek = element_type(inputs[k])` is sufficient to describe all the correspondence relationships.